uuid_source.go

```diff
@@ -0,0 +1,54 @@
+package uuid
+
+import (
+	"io"
+	"crypto/rand"
+)
+
+// A UuidSource holds a random number generator and generates UUIDs using it as its source.
+//
+// It is useful when a process need its own random number generator, 
+// e.g. when running some processes concurrently.
+type UuidSource struct {
```

Comment:
It would be UUIDSource, but if you are mimicking math/rand then just call it Source.

One issue with this setup it you still can only have a single rander function.  Rather than add all of this complexity, it might be easier to simply create one new function, say NewRandomFromReader (I don't really like that name, it is a bit of a mouthful, but I don't expect it to be used often at all).

```
func NewRandomFromReadr(r io.Reader) (UUID, error)
```

Now an external package can create this type if they want.

```
type Source struct {
        r io.Reader
}

func NewUUIDSource(r io.Reader) Source {
        return Source{r: r}
}

func (s Source) NewRandom() (uuid.UUID, error) {
        return uuid.NewRandomFromReader(s.r)
}
```

That will keep the uuid package cleaner.

Reply:
@pborman , Thanks for looking into this PR!
Let me ask to clarify - are you suggesting here that instead of everything in the uuid_source file, we just add the `func NewRandomFromReadr(r io.Reader) (UUID, error)` function to `version4.go` and let the user build the structs around it if they need? I am not clear, as you are also commenting on the content of the uuid_source files.

Reply:
Yes, that is exactly correct.  And yes, the NewRandomFromReader is essentially the newRandom function you added.  The uuid_source files would no longer be needed, although a simple test showing that NewRandomFromReader did in fact use the supplied reader would probably be a good addition.  This could be very simple where you have a bytes.NewReader that has 16 bytes of data and the first NewRandomFromReader gives you the UUID you expect and the second NewRandomFromReader returns an error.

You might want to consider putting a package in GitHub that has the functionality of your uuid_source file so others can also use it if they need that sort of functionality.

Reply:
ok, Got it. 

> You might want to consider putting a package in GitHub that has the functionality of your uuid_source file so others can also use it if they need that sort of functionality.

A package for using it seems too much for just few lines of code. Maybe I'll just add an example?

Reply:
@pborman